runtime/acala/src/lib.rs

```diff
@@ -1469,6 +1469,33 @@ impl<I: From<Balance>> frame_support::traits::Get<I> for TxFeePerGasV2 {
 	}
 }
 
+pub struct RandomnessSource<T>(sp_std::marker::PhantomData<T>);
+impl<T: frame_system::Config> Randomness<T::Hash, BlockNumber> for RandomnessSource<T> {
+	fn random(subject: &[u8]) -> (T::Hash, BlockNumber) {
+		let relay_storage_root = ParachainSystem::validation_data()
+			.expect("set in `set_validation_data`")
```

Comment:
ValidationData is removed at `on_initialize` and set at the inherent, this means it could be empty in the `on_initialize` hook for some pallets and some other inherents
so this could fail when invoked by scheduler or some other pallet's `on_initialize` hook

Reply:
we better to avoid expect here, when no random value is available, we can return an insecure random value and block number of zero